### PR TITLE
[REEF-893] Different communication groups should use different thread…

### PR DESCRIPTION
**Pull request by gyeongin**
…s to handle group comm messages

This addressed the issue by
  * Change `groupCommMessageStage` field in `GroupCommDriverImpl` to use `SyncStage`
  * Each `CommunicationGroupDriverImpl` uses `SingleThreadStage` for message handling
  * Add test to examine corner case which fails with current implementation

JIRA:
  [REEF-JIRA_ISSUE_NUMBER](https://issues.apache.org/jira/browse/REEF-893)

### CODE REVIEW

File: lang/java/reef-tests/src/main/java/org/apache/reef/tests/group/MasterTask.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+package org.apache.reef.tests.group;
+
+import org.apache.reef.io.network.group.api.task.CommunicationGroupClient;
+import org.apache.reef.io.network.group.api.task.GroupCommClient;
+import org.apache.reef.task.Task;
+
+import javax.inject.Inject;
+
+/**
+ * Master task used for testing multiple communication groups
```
Comment (dongjoon-hyun) [inline lang/java/reef-tests/src/main/java/org/apache/reef/tests/group/MasterTask.java]: Please add '.' at the end.

File: lang/java/reef-tests/src/main/java/org/apache/reef/tests/group/SlaveTask.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+package org.apache.reef.tests.group;
+
+import org.apache.reef.io.network.group.api.task.CommunicationGroupClient;
+import org.apache.reef.io.network.group.api.task.GroupCommClient;
+import org.apache.reef.task.Task;
+
+import javax.inject.Inject;
+
+/**
+ * Slave task used for testing multiple communication groups
```
Comment (dongjoon-hyun) [inline lang/java/reef-tests/src/main/java/org/apache/reef/tests/group/SlaveTask.java]: Please add '.' at the end.

**Comment by dongjoon-hyun** [comment]
Hi, @gyeongin . It's checkstyle failures. I leave a few comments.

**Comment by gyeongin** [comment]
Thanks, @dongjoon-hyun, I missed it.

**Comment by dongjoon-hyun** [comment]
Now, I will do another pass, @gyeongin .

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/CommunicationGroupDriverFactory.java [hunk 3]
```diff
@@ -53,16 +53,19 @@ private CommunicationGroupDriverFactory(
...
     injector.bindVolatileParameter(GroupCommSenderStage.class, senderStage);
     injector.bindVolatileParameter(DriverIdentifier.class, driverId);
     injector.bindVolatileParameter(GroupCommRunningTaskHandler.class, groupCommRunningTaskHandler);
     injector.bindVolatileParameter(GroupCommFailedTaskHandler.class, groupCommFailedTaskHandler);
     injector.bindVolatileParameter(GroupCommFailedEvalHandler.class, groupCommFailedEvaluatorHandler);
+    injector.bindVolatileInstance(GroupCommMessageHandler.class, groupCommMessageHandler);
   }
 
   /**
+   * Deprecated in 0.14.
```
Comment (dongjoon-hyun) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/CommunicationGroupDriverFactory.java]: Please use `@deprecated`.

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommMessageHandler.java [hunk 4]
```diff
@@ -35,13 +38,24 @@
 
   private static final Logger LOG = Logger.getLogger(GroupCommMessageHandler.class.getName());
 
-  private final Map<Class<? extends Name<String>>, BroadcastingEventHandler<GroupCommunicationMessage>>
-      commGroupMessageHandlers = new HashMap<>();
+  private final Map<Class<? extends Name<String>>, AbstractEStage<GroupCommunicationMessage>>
+      commGroupMessageStages = new HashMap<>();
 
+  /**
+   * Deprecated in 0.14.
```
Comment (dongjoon-hyun) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommMessageHandler.java]: Please use `@deprecated`.

**Comment by dongjoon-hyun** [comment]
It looks good to me. I will fix above two comments and merge this.
